File: example-java/app/src/main/java/com/vibes/androidsdkexampleapp/model/MessagesAdapter.java
package com.vibes.androidsdkexampleapp.model;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.squareup.picasso.Picasso;
import com.vibes.androidsdkexampleapp.R;
import com.vibes.vibes.InboxMessage;

import java.text.DateFormat;
import java.util.List;

public class MessagesAdapter extends RecyclerView.Adapter<MessagesAdapter.ViewHolder> {
    private List<InboxMessage> messages;
    private LayoutInflater layoutInflater;
    private InboxMessageClickListener inboxMessageClickListener;
    private Context context;

    public MessagesAdapter(Context context, List<InboxMessage> messages) {
        this.messages = messages;
        this.context = context;
        this.layoutInflater = LayoutInflater.from(context);
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = layoutInflater.inflate(R.layout.inbox_messages_row, parent, false);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        InboxMessage inboxMessage = messages.get(position);
        holder.subject.setText(inboxMessage.getSubject());
        holder.content.setText(inboxMessage.getContent());
        Picasso.get().load(inboxMessage.getIconImage()).into(holder.iconImage);
        DateFormat dateFormat = android.text.format.DateFormat.getDateFormat(context);
        holder.expiresAt.setText(dateFormat.format(inboxMessage.getExpirationDate()));
    }

    @Override
    public int getItemCount() {
        return messages.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        TextView subject;
        TextView content;
        TextView expiresAt;
        ImageView iconImage;
        View view;
        ViewHolder(View itemView) {
            super(itemView);
            view = itemView;
            subject = view.findViewById(R.id.subject);
            content = view.findViewById(R.id.content);
            iconImage = view.findViewById(R.id.imageIconView);
            expiresAt = view.findViewById(R.id.expires_at);
            view.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            if(inboxMessageClickListener != null) inboxMessageClickListener.onInboxMessageClick(view, getAdapterPosition());
        }
    }

    public InboxMessage getMessage(int id) {
        return messages.get(id);
    }

    public void setInboxMessageClickListener(InboxMessageClickListener inboxMessageClickListener) {
        this.inboxMessageClickListener = inboxMessageClickListener;
    }
    public interface InboxMessageClickListener {
        void onInboxMessageClick(View view, int position);
    }
}

File: example-kotlin/app/src/main/java/com/android/exampleapp/kotlin/adapters/MessagesAdapter.kt
package com.android.exampleapp.kotlin.adapters

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.text.format.DateFormat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.android.exampleapp.kotlin.R
import com.squareup.picasso.Picasso
import com.vibes.vibes.InboxMessage

class MessagesAdapter(context: Context?, messages: List<InboxMessage?>) : RecyclerView.Adapter<MessagesAdapter.ViewHolder>() {
    private var messages: List<InboxMessage?>? = null
    private var layoutInflater: LayoutInflater? = null
    private var context: Context? = null
    var onItemClick: ((InboxMessage?) -> Unit)? = null

    init {
        this.messages = messages
        this.context = context
        layoutInflater = LayoutInflater.from(context)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemView: View = layoutInflater!!.inflate(R.layout.inbox_messages_row, parent, false)
        return ViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val inboxMessage = messages!![position]
        holder.subject.text = inboxMessage!!.subject
        holder.content.text = inboxMessage.content
        Picasso.get().load(inboxMessage.iconImage).into(holder.iconImage)
        val dateFormat = DateFormat.getDateFormat(context)
        holder.expiresAt.text = dateFormat.format(inboxMessage.expirationDate)
    }

    override fun getItemCount(): Int {
        return messages!!.size
    }

    inner class ViewHolder internal constructor(view: View) : RecyclerView.ViewHolder(view) {
        var subject: TextView = view.findViewById(R.id.subject)
        var content: TextView = view.findViewById(R.id.content)
        var expiresAt: TextView = view.findViewById(R.id.expires_at)
        var iconImage: ImageView = view.findViewById(R.id.imageIconView)

        init {
            view.setOnClickListener { onItemClick?.invoke(messages?.get(adapterPosition)) }
        }
    }
}
